cachelib/navy/common/FdpNvme.cpp

```diff
@@ -0,0 +1,321 @@
+  size_t pPos = blockDevice.find('p', bPos);
+
+  XDCHECK(bPos < std::string::npos);
+  return blockDevice.substr(bPos, pPos - bPos);
+}
+
+// It returns nvme0n1p1 for /dev/nvme0n1p1 (partitioned NS),
+// and returns "" for /dev/nvme0n1 (non-partitioned NS)
+std::string getNvmePartName(const std::string& blockDevice) {
+  size_t bPos = blockDevice.find_last_of("/") + 1;
+  size_t pPos = blockDevice.find('p', bPos);
+
+  XDCHECK(bPos < std::string::npos);
+  if (pPos == std::string::npos) {
+    return {}; // The block device is just an NS; not a partition
+  }
+  return blockDevice.substr(bPos, std::string::npos - bPos);
+}
+
+// Read the /sys/block/xx entry for a block device
+std::string readDevAttr(const std::string& bName, const std::string& attr) {
+  std::string path = "/sys/block/" + bName + '/' + attr;
+  std::string entry;
+  if (!folly::readFile(path.c_str(), entry)) {
+    throw std::runtime_error(folly::sformat("Unable to read {}", path));
+  }
+  return entry;
+}
+
+// Get the Namespace ID of a NVMe block device
+int getNvmeNsId(const std::string& bName) {
+  return (folly::to<int>(readDevAttr(bName, "nsid")));
+}
+
+// Get the Max Transfer size in bytes for an NVMe block device
+uint32_t getMaxTfrSize(const std::string& bName) {
+  // max_hw_sectors_kb : This is the maximum number of kilobytes supported in a
+  // single data transfer.
+  // (https://www.kernel.org/doc/Documentation/block/queue-sysfs.txt)
+  return (1024u * /* multiply by kb */
+          folly::to<uint32_t>(readDevAttr(bName, "queue/max_hw_sectors_kb")));
+}
+
+// Get LBA shift of an NVMe block device
+uint32_t getLbaShift(const std::string& bName) {
+  return folly::constexpr_log2(
+      folly::to<uint32_t>(readDevAttr(bName, "queue/logical_block_size")));
+}
+
+// Get the partition start in bytes
+uint64_t getPartStart(const std::string& bName, const std::string& pName) {
+  return (512u * /* sysfs size is in terms of linux sector size */
+          folly::to<uint64_t>(readDevAttr(bName + "/" + pName, "start")));
+}
+
+NvmeData FdpNvme::readNvmeInfo(const std::string& blockDevice) {
+  auto bName = getNvmeBlockName(blockDevice);
+  auto pName = getNvmePartName(blockDevice);
+
+  try {
+    int namespace_id = getNvmeNsId(bName);
+    uint32_t lbaShift = getLbaShift(bName);
+    uint32_t maxTfrSize = getMaxTfrSize(bName);
+
+    uint64_t startLba{0};
+    if (!pName.empty()) {
+      startLba = getPartStart(bName, pName) >> lbaShift;
+    }
+
+    XLOGF(INFO,
+          "Nvme Device Info, NS Id: {}, lbaShift: {},"
+          " Max Transfer size: {}, start Lba: {}",
+          namespace_id, lbaShift, maxTfrSize, startLba);
+
+    return NvmeData{namespace_id, lbaShift, maxTfrSize, startLba};
+  } catch (const std::system_error& e) {
+    XLOGF(ERR, "Exception in openNvmeCharFile for: {}", blockDevice);
+    throw;
+  }
+}
+
+bool isValidNvmeDevice(const std::string& fileName) {
+  return std::regex_match(fileName, std::regex("^/dev/nvme\\d+n\\d+(p\\d+)?$"));
```

Comment:
Could you replace this with the equivalent (`re2::RE2::FullMatch`?) from RE2? https://github.com/google/re2 

We cannot use `regex_match` due to its vulnerability to segfault https://gcc.gnu.org/bugzilla/show_bug.cgi?id=61582&fbclid=IwAR16Xwiru8QryYObTnoATM5WwCSTX9IRVorMwDxYq9E0sxrsBQTw-H1iKH4